File: typescript/libs/workspace/domain/system/src/lib/runtime/ISession.ts
import { Class, Composite } from '@allors/workspace/meta/system';
import { IObject } from './IObject';
import { IWorkspace } from './IWorkspace';
import { IChangeSet } from './IChangeSet';
import { InvokeOptions } from '../api/pull/InvokeOptions';
import { IInvokeResult } from '../api/pull/IInvokeResult';
import { Procedure } from '../api/pull/Procedure';
import { Pull } from '../api/pull/Pull';
import { IPullResult } from '../api/pull/IPullResult';
import { IPushResult } from '../api/push/IPushResult';
import { ISessionServices } from './ISessionServices';
import { Method } from './Method';
import { IWorkspaceResult } from './IWorkspaceResult';

export interface ISession {
  workspace: IWorkspace;

  services: ISessionServices;

  pullFromWorkspace(): IWorkspaceResult;

  pushToWorkspace(): IWorkspaceResult;

  checkpoint(): IChangeSet;

  create<T extends IObject>(cls: Class): T;

  instantiate<T extends IObject>(id: number): T;
  instantiate<T extends IObject>(ids: number[]): T[];
  instantiate<T extends IObject>(objectType: Composite): T[];
}

File: dotnet/System/workspace/csharp/allors.workspace/domain/runtime/ISession.cs
// <copyright file="IDatabase.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using System.Threading.Tasks;
    using Data;
    using Meta;
    using System.Collections.Generic;

    public interface ISession
    {
        IWorkspace Workspace { get; }

        ISessionServices Services { get; }

        Task<IInvokeResult> Invoke(Method method, InvokeOptions options = null);

        Task<IInvokeResult> Invoke(Method[] methods, InvokeOptions options = null);

        Task<IPullResult> Call(Procedure procedure, params Pull[] pull);

        Task<IPullResult> Pull(params Pull[] pull);

        Task<IPushResult> Push();

        IWorkspaceResult PullFromWorkspace();

        IWorkspaceResult PushToWorkspace();

        IChangeSet Checkpoint();

        T Create<T>() where T : class, IObject;

        T Create<T>(IClass @class) where T : class, IObject;

        #region Instantiate
        T Instantiate<T>(IObject @object) where T : class, IObject;

        T Instantiate<T>(T @object) where T : class, IObject;

        T Instantiate<T>(long? id) where T : class, IObject;

        T Instantiate<T>(long id) where T : class, IObject;

        T Instantiate<T>(string idAsString) where T : class, IObject;

        IEnumerable<T> Instantiate<T>(IEnumerable<IObject> objects) where T : class, IObject;

        IEnumerable<T> Instantiate<T>(IEnumerable<T> objects) where T : class, IObject;

        IEnumerable<T> Instantiate<T>(IEnumerable<long> ids) where T : class, IObject;

        IEnumerable<T> Instantiate<T>(IEnumerable<string> ids) where T : class, IObject;

        IEnumerable<T> Instantiate<T>() where T : class, IObject;

        IEnumerable<T> Instantiate<T>(IComposite objectType) where T : class, IObject;
        #endregion
    }
}
